File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/MapboxMapFactory.kt
package com.mapbox.maps.mapbox_maps

import android.annotation.SuppressLint
import android.content.Context
import com.mapbox.common.FeatureTelemetryCounter
import com.mapbox.maps.MapInitOptions
import com.mapbox.maps.MapOptions
import com.mapbox.maps.applyDefaultParams
import com.mapbox.maps.mapbox_maps.pigeons._MapInterface
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class MapboxMapFactory(
  private val messenger: BinaryMessenger,
  private val flutterAssets: FlutterPlugin.FlutterAssets,
  private val lifecycleProvider: MapboxMapsPlugin.LifecycleProvider
) : PlatformViewFactory(_MapInterface.codec) {

  @SuppressLint("RestrictedApi")
  override fun create(context: Context?, viewId: Int, args: Any?): PlatformView {
    if (context == null) {
      throw RuntimeException("Context is null, can't create MapView!")
    }
    val params = args as Map<String, Any>
    val mapOptions = params["mapOptions"] as com.mapbox.maps.mapbox_maps.pigeons.MapOptions?
    val cameraOptions = params["cameraOptions"] as com.mapbox.maps.mapbox_maps.pigeons.CameraOptions?
    val channelSuffix = params["channelSuffix"] as Long
    val textureView = params["textureView"] as? Boolean ?: false
    val styleUri = params["styleUri"] as? String
    val styleJson = params["styleJson"] as? String
    val initialScaleBarEnabled = params["initialScaleBarEnabled"] as? Boolean
    val initialCompassEnabled = params["initialCompassEnabled"] as? Boolean
    val pluginVersion = params["mapboxPluginVersion"] as String
    val eventTypes = params["eventTypes"] as List<Long>

    // When styleJson is provided it takes precedence — we load it post-init
    // via MapboxMap.loadStyleJson. MapInitOptions.styleUri stays null so the
    // SDK doesn't start a redundant URI load first.
    val mapInitOptions = MapInitOptions(
      context = context,
      mapOptions = mapOptions?.toMapOptions(context) ?: MapOptions.Builder()
        .applyDefaultParams(context).build(),
      cameraOptions = cameraOptions?.toCameraOptions(context),
      textureView = textureView,
      styleUri = if (styleJson != null) null else styleUri
    )
    mapCounter.increment()
    return MapboxMapController(
      context,
      mapInitOptions,
      lifecycleProvider,
      messenger,
      channelSuffix,
      pluginVersion,
      eventTypes,
      initialScaleBarEnabled,
      initialCompassEnabled,
      styleJson
    )
  }

  companion object {
    @SuppressLint("RestrictedApi")
    private val mapCounter = FeatureTelemetryCounter.create("maps-mobile/flutter/map")
  }
}

File: ios/mapbox_maps_flutter/Sources/mapbox_maps_flutter/Classes/MapboxMapFactory.swift
import Flutter
import MapboxMaps
import MapboxCommon
import MapboxCommon_Private

final class MapboxMapFactory: NSObject, FlutterPlatformViewFactory {
    private static let mapCounter = FeatureTelemetryCounter.create(forName: "maps-mobile/flutter/map")

    var registrar: FlutterPluginRegistrar

    deinit {
        _MapboxOptionsSetup.setUp(binaryMessenger: registrar.messenger(), api: nil)
        _MapboxMapsOptionsSetup.setUp(binaryMessenger: registrar.messenger(), api: nil)
    }

    init(withRegistrar registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return MapInterfacesPigeonCodec.shared
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {

        guard let args = args as? [String: Any] else {
            return MapboxMapController(
                withFrame: frame,
                mapInitOptions: MapInitOptions(),
                channelSuffix: 0,
                registrar: registrar,
                pluginVersion: "",
                eventTypes: []
            )
        }

        let styleJson = args["styleJson"] as? String
        // styleJson takes precedence; skip the URI load when both are set.
        let styleURI = styleJson == nil
            ? (args["styleUri"] as? String).flatMap(StyleURI.init(rawValue:))
            : nil
        let mapOptions = args["mapOptions"] as? MapOptions
        let cameraOptions = args["cameraOptions"] as? CameraOptions
        let initialScaleBarEnabled = args["initialScaleBarEnabled"] as? Bool
        let initialCompassEnabled = args["initialCompassEnabled"] as? Bool

        let mapInitOptions = MapInitOptions(
            mapOptions: mapOptions?.toMapOptions() ?? MapboxMaps.MapOptions(),
            cameraOptions: cameraOptions?.toCameraOptions(),
            styleURI: styleURI
        )

        Self.mapCounter.increment()
        return MapboxMapController(
            withFrame: frame,
            mapInitOptions: mapInitOptions,
            channelSuffix: args["channelSuffix"] as? Int ?? 0,
            registrar: registrar,
            pluginVersion: args["mapboxPluginVersion"] as? String ?? "",
            eventTypes: args["eventTypes"] as? [Int] ?? [],
            initialScaleBarEnabled: initialScaleBarEnabled,
            initialCompassEnabled: initialCompassEnabled,
            styleJson: styleJson
        )
    }
}
